New features and bug fixes

Madu86 opened the pull request:
This merge contains new features: configure script based build system, API implementation and a cudampi version. Furthermore, a series of memory leaks and bugs have been fixed. Some parts of the legacy code have also been refactored.

agoetz commented on configure:
```diff
@@ -0,0 +1,625 @@
+    mkdir -p "$quick_prefix/test"
+    echo "Creating symlink for $quick_prefix/test"
+    cd "$quick_prefix/test"
+    lndir "$QUICK_HOME/test"
```
`lndir` is by default not available on CentOS. We should not rely on commands that must be installed with user packages.

Madu86 replied:
I have fixed this in commit fd802a5eae249b80c9852d3ef284124277225cf9.

agoetz replied:
This is fixed. Resolving conversation.

agoetz commented on configure:
```diff
@@ -0,0 +1,999 @@
+ctime=`date +'%r'`
+
+echo ""
+echo -e "\033[92mConfiguring $QUICK_VERSION installation started on $cdate at $ctime..\033[0m" 
```
When using `/bin/sh` on a MacOS (I have Mojave), the `echo` command literally prints `-e`:
```
andi@blitz:~/projects/quick/quick>./configure
-e Configuring QUICK-20.06 installation started on 07/29/2020 at 01:11:36 AM..
-e                                                                                                                                        
  Use this script as ./configure [flags] compiler                           
```
Also, can we keep it default color (i.e. black for me)? The neon green looks quite harsh on my eyes, in particular on white background. Darker green (color 32) might work. But then again, it will depend on the color people have in their terminal. I'd vouch for not using colot.

Madu86 replied:
Sure. I can remove all the colors.

agoetz commented:
When running `./configure -serial gnu` on MacOS, I get
```
...
  gcc version 8.4.0 found..
grep: repetition-operator operand invalid
  g++ version 8.4.0 found..
```
I haven't quite figured out which grep has a non-standard expression. Any idea?

Madu86 commented:
> When running `./configure -serial gnu` on MacOS, I get
> 
> ```
> ...
>   gcc version 8.4.0 found..
> grep: repetition-operator operand invalid
>   g++ version 8.4.0 found..
> ```
> 
> I haven't quite figured out which grep has a non-standard expression. Any idea?

Yeah, https://github.com/Madu86/QUICK/blob/master/configure#L272. Should be fixed as  `if [  "$os_name" = 'osx' ] && [ -z "$cc_version" ]; then`

agoetz commented:
> > When running `./configure -serial gnu` on MacOS, I get
> > ```
> > ...
> >   gcc version 8.4.0 found..
> > grep: repetition-operator operand invalid
> >   g++ version 8.4.0 found..
> > ```
> > 
> > 
> > I haven't quite figured out which grep has a non-standard expression. Any idea?
> 
> Yeah, https://github.com/Madu86/QUICK/blob/master/configure#L272. Should be fixed as `if [ "$os_name" = 'osx' ] && [ -z "$cc_version" ]; then`

This does not fix it. I still get `grep: repetition-operator operand invalid`.

agoetz commented:
I have been playing with the `configure` script. A few notes, in no particular order
- We should use `--` for multi-letter flags. By convention only single letter flags use a single dash `-`, i.e. `-serial` should be interpreted as `-s -e -r -i -a -l`. Instead, we should use `--serial`. Amber's configure is not consistent but we should not copy that.
- `configure` only checks serial compilers
- Not specifying a build type flag (`./configure gnu`) results in empty `BUILDTYPES`. Should default to serial.
- It is possible to specify `-arch` without `-cuda`. This results in an empty `BUILDTYPES`
- Cannot specify two architectures as either `-arch pascal volta` or `-arch pascal -arch volta`
- Shared objects are not built. `./configure -serial -shared gnu` will build `build/serial/lib/*.a`, no `*.so` files.
- Related to previous point, this results in link errors: `./configure -serial gnu; make; ./configure -shared -serial gnu; make`.
- It would be good to have all build settings in a single configuration file. Instead of generating `make.in` we could generate `build/make.in.serial`, `build/make.in.mpi` etc and include these in the master `Makefile` and in the architecture specific Makefiles in `build/serial/Makefile` etc. This would make it easy to check and manually change settings.
- `make clean` removes not only the object files but also the binaries and library files. (configured w/o prefix)
- Cannot use `make -j`. This is standard for parallel make. It's quite weird to specify number of cpus during configure. I've never seen this before.
We don't have to fix these points now. I just would like to keep track. I think we should start using the Github Projects feature. E.g. have a Project "Build System". Another project "CMake Build System" etc.

agoetz commented:
Running tests is much more convenient now. We can probably improve it, for instance to be able to say `make test.serial`. Currently tests are run for all executables that are found. For a developer I may want to keep serial and MPI versions but make changes only to CUDA code, hence test only the CUDA executable. It would also be nice not to have the `runtest` script copied. It's not something we have to look into now but can keep in mind.

agoetz commented:
`make clean` does not remove files in `build/common/gpu/libxcdevobj/`.

agoetz commented:
If specifying an installation prefix via `./configure --prefix installdir`
- how do we run tests from the install directory?
- we cannot move or remove the source directory because of symbolic links for the tests

agoetz commented:
On MacOS, `make` somehow triggers a build and an install, while `make install` results in 
```
usage: cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file target_file
       cp [-R [-H | -L | -P]] [-fi | -n] [-apvXc] source_file ... target_directory
make: *** [serialinstall] Error 64
```

agoetz commented on examples/gly12_hf_energy.in:
```diff
@@ -1,91 +0,0 @@
- HF BASIS=cc-pvDZ CUTOFF=1.0d-10 DENSERMS=1.0d-8 ENERGY
```
File name: Add also basis set. `gly12_hf_cc-pvdz_energy.in`.

Madu86 replied:
We no longer have an examples folder in QUICK. I plan to maintain a collection of examples in a separate repository.

agoetz commented on src/cuda/Makefile:
```diff
@@ -1,111 +1,105 @@
-	
+	@-rm -f *.o 
+	@-rm -f ./CUBLAS/*.o
+	@-rm -f ./CUSOLVER/*.o
```
These folders don't exist. I see folders with lower-case letters. 
I also do not understand why we need the files in folders `CUBLAS` and `CUSOLVER`. Shouldn't these be part of the CUDA Toolkit?

Madu86 replied:
This has been fixed. What we have in cublas and cusolver folders are wrapper functions.

agoetz commented on src/cuda/gpu_common.h:
```diff
@@ -130,15 +134,14 @@ cudaEventDestroy(end);
+fprintf(gpu->debugFile,"METHOD = %i\n", gpu->gpu_sim.method);\
 x;\
 TIMERSTOP()\
 cudaError_t error = cudaGetLastError();\
```
I think we should keep error checking also in the production code. And perhaps the timer information although I don't know if it is used.

Madu86 replied:
Lets create a new project for QUICK error handling.

agoetz commented on src/libxc/lda_xc_1d_ehwlrg.c:
```diff
@@ -26,7 +26,7 @@ lda_xc_1d_ehwlrg_init(xc_func_type *p)
   assert(p!=NULL && p->params == NULL); 
   p->params = malloc(sizeof(lda_xc_1d_ehwlrg_params)); 
-#ifdef CUDA 
+#if defined CUDA || defined CUDA_MPIV 
```
Is this really necessary? Don't we always have CUDA=True if CUDA_MPIV=True? In other words, a CUDA MPI build could just set -DCUDA. Am I missing something? We don't have to change it, I am just wondering.

agoetz replied:
Ignore this comment.

agoetz commented on test/ene_psb3_libxc_gga_631g.in:
```diff
@@ -0,0 +1,18 @@
+ DFT LIBXC=LIBXC=GGA_X_B88,GGA_C_LYP BASIS=6-31G CUTOFF=1.0d-10 DENSERMS=1.0d-6 CHARGE=+1
```
Looks like an extra `LIBXC=`

Madu86 replied:
Fixed!

agoetz commented on tools/build-makefile:
```diff
@@ -0,0 +1,224 @@
+#  !---------------------------------------------------------------------!
```
It seems dangerous to have build definitions separated from the source tree.

agoetz replied:
I also don't fully understand this but it seems to duplicate at least part of the information in `src/libxc/Makefile` and `src/cuda/Makefile`.

agoetz commented:
Compilation for Pascal architecture fails. I am on chinotto. Have you seen this?
```
./configure -serial -mpi -cuda gnu
make
...
...
nvlink error   : Entry function '_Z20getxc_kernel_new_impPP14gpu_libxc_infoi' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants (target: sm_60)
...
```

agoetz commented on src/schwarz.f90:
```diff
@@ -18,7 +17,7 @@ subroutine schwarzoff
 #endif
 
-  integer ii,jj
+  integer i,ii,jj
```
integer i is not used.

Madu86 replied:
Fixed.

agoetz commented:
Maybe `runtest` should be copied into the install directory if a prefix is given. Or the install/test directory. If `runtest` accepts a buildtype as flag, e.g. `runtest serial` etc. then we can also have test targets in the Makefile to invoke `make test.serial` etc.

agoetz commented on src/quick_one_electron_integral.f90:
```diff
@@ -124,51 +123,51 @@ subroutine fullx
-   ekinetic = kinetic/(-2.d0)
-   return
-end function ekinetic
+!double precision function ekinetic(a,b,i,j,k,ii,jj,kk,Ax,Ay,Az,Bx,By,Bz)
```
The code is identical to `subs/ekinetic.f90` but this version is better because it got rid of `implicit double precision (a-h, o-z)` and is nicer formatted. We should keep this version.

Madu86 replied:
I agree. We should create a new project for refactoring legacy code and put this in there.

agoetz commented on src/quick_one_electron_integral.f90:
```diff
@@ -1751,68 +1750,65 @@ subroutine get1e(oneElecO)
-      !------- END MPI/ALL NODES ------------
-   endif
-#endif
+!#ifdef MPIV
```
Why are we getting rid of MPI for the 1e integrals?

Madu86 replied:
It doesnt work and we never used it. Have to reimplement the whole one electron code.

agoetz commented on src/modules/quick_api_module.f90:
```diff
@@ -0,0 +1,688 @@
+    double precision, allocatable, dimension(:,:) :: coords
+
+    ! charge and the coordinates of external point charges
+    ! size is 4*nxt_ptchg, where first element of a row holds the charge
```
Did we not change this to store (x, y, z, q)?

Madu86 replied:
We did. Just corrected the comment.

agoetz commented on src/modules/quick_api_module.f90:
```diff
@@ -0,0 +1,688 @@
+
+    ! charge and the coordinates of external point charges
+    ! size is 4*nxt_ptchg, where first element of a row holds the charge
+    double precision, allocatable, dimension(:,:) :: xt_chg_crd
```
Maybe be consistent in naming. Here you use xt_chg, below ptchg

Madu86 replied:
Fixed.

agoetz commented on src/modules/quick_api_module.f90:
```diff
@@ -0,0 +1,688 @@
+
+  ! we will reuse density matrix for steps above 1. For the 1st step, we should
+  ! read basis file and run SAD guess. 
+  if(self%firstStep .and. self%reuse_dmx) then
```
This seems wrong. Should this have an `or` statement? Since if we don't reuse the density matrix, then we will need to do a new guess.

Madu86 replied:
No, I just want to reuse the density matrix computed from the first step. For the first step, self%firstStep and self%reuse_dmx are both true. We will compute density matrix for this case, save it and set the value of self%firstStep to .false. In the next step, we will reuse the computed density matrix and code paths within the if statement will not be executed.

agoetz replied:
I still don't understand. At the first step we need to do a guess because we have no density matrix. At subsequent steps we can choose to do a new guess or reuse the density matrix from the previous step. I have to admit that I don't know what happens if `quick_method%SAD` is `.false.`.  I also assume that we need to assign basis functions only during first step. Then we need
```
if (self%firstStep) then
  call getMol()
end if

if (self%firstStep .or. (.not. self%reuse_dmx)) then
  if (quick_method%SAD) call getMolSad()
  ! What do we do if we don't do a SAD guess? 
  ! Where do we delete the density matrix?
end if

self%firstStep = .false.
```

agoetz commented on src/modules/quick_api_module.f90:
```diff
@@ -0,0 +1,688 @@
+      endif
+#endif
+
+  if (failed) call quick_exit(iOutFile,1)
```
We probably should communicate a failure back to the MD code.

Madu86 replied:
We can add this task to error handling project.

agoetz commented on src/modules/quick_api_module.f90:
```diff
@@ -0,0 +1,688 @@
+
+! convert gradients into gradients and save in quick_api struct &
+! save point charge gradients too. Note that quick_qm_struct saves
+! both gradients in vector formats and we should organize them
```
In principle the layout in memory should be the same since Fortran stores in column major format, i.e. we have x1, y1, z1, x2, y2, z2, ... both in quick_api struct and quick_qm_struct. But being explicit is probably better.

Madu86 replied:
I agree. I also corrected the comment.

Madu86 commented:
> I have been playing with the `configure` script. A few notes, in no particular order
> 
> * We should use `--` for multi-letter flags. By convention only single letter flags use a single dash `-`, i.e. `-serial` should be interpreted as `-s -e -r -i -a -l`. Instead, we should use `--serial`. Amber's configure is not consistent but we should not copy that.

Fixed. We use double dashes now. 

> * `configure` only checks serial compilers

This is because mpi wrapper compilers spit out their base compiler names with -v flag. For instance,  `mpiifort -v ` results in  "ifort version 18.0.5". This is what I print. 

> * Not specifying a build type flag (`./configure gnu`) results in empty `BUILDTYPES`. Should default to serial.

Fixed.

> * It is possible to specify `-arch` without `-cuda`. This results in an empty `BUILDTYPES`

Fixed. We will now silently ignore this flag if cuda/cudampi is not specified. 

> * Cannot specify two architectures as either `-arch pascal volta` or `-arch pascal -arch volta`

Fixed. You can now use `--arch volta --arch pascal` to compile for both archs.

> * Shared objects are not built. `./configure -serial -shared gnu` will build `build/serial/lib/*.a`, no `*.so` files.

> * Related to previous point, this results in link errors: `./configure -serial gnu; make; ./configure -shared -serial gnu; make`.
> * Related to previous point, there is no obvious information to see that we configured for shared libraries. See also next point.

Fixed.

> * It would be good to have all build settings in a single configuration file. Instead of generating `make.in` we could generate `build/make.in.serial`, `build/make.in.mpi` etc and include these in the master `Makefile` and in the architecture specific Makefiles in `build/serial/Makefile` etc. This would make it easy to check and manually change settings.

Skipped for now. 

> * `make clean` removes not only the object files but also the binaries and library files. (configured w/o prefix)

Fixed. 

> * Cannot use `make -j`. This is standard for parallel make. It's quite weird to specify number of cpus during configure. I've never seen this before.

We have a circular module dependency in legacy code. Cannot directly use `make -j` until we fix that.

> 
> We don't have to fix these points now. I just would like to keep track. I think we should start using the Github Projects feature. E.g. have a Project "Build System". Another project "CMake Build System" etc.

Madu86 commented:
> Compilation for Pascal architecture fails. I am on chinotto. Have you seen this?
> 
> ```
> ./configure -serial -mpi -cuda gnu
> make
> ...
> ...
> nvlink error   : Entry function '_Z20getxc_kernel_new_impPP14gpu_libxc_infoi' uses too much data for compiler-generated constants; please recompile with -Xptxas --disable-optimizer-constants (target: sm_60)
> ...
> ```

It seems that device code linking has some limitations on pascal architecture. I temporarily fixed the problem by introducing a variable for `-Xptxas --disable-optimizer-constants` flag for compiling gpu_getxc.cu and libxc device code. But this leads to very slow XC calculations, at least on GTX1080.

Madu86 commented:
> Running tests is much more convenient now. We can probably improve it, for instance to be able to say `make test.serial`. Currently tests are run for all executables that are found. For a developer I may want to keep serial and MPI versions but make changes only to CUDA code, hence test only the CUDA executable. It would also be nice not to have the `runtest` script copied. It's not something we have to look into now but can keep in mind.

I enabled user specified test runs in the test script.

Madu86 commented:
@agoetz Andy,  I have addressed most of your suggestions.  Please review the pull request again and let me know if you find other things that we should fix.

agoetz commented:
> > > When running `./configure -serial gnu` on MacOS, I get
> > > ```
> > > ...
> > >   gcc version 8.4.0 found..
> > > grep: repetition-operator operand invalid
> > >   g++ version 8.4.0 found..
> > > ```
> > > 
> > > 
> > > I haven't quite figured out which grep has a non-standard expression. Any idea?
> > 
> > 
> > Yeah, https://github.com/Madu86/QUICK/blob/master/configure#L272. Should be fixed as `if [ "$os_name" = 'osx' ] && [ -z "$cc_version" ]; then`
> 
> This does not fix it. I still get `grep: repetition-operator operand invalid`.

It's in lines 288 and 291, though I don't know what is the cause:
```
cxx_version_info=`$cxx_search -v 2>&1 | grep -oE "$cxx_search\s[vV]ersion\s[0-9]{1,4}.[0-9]{1,4}.[0-9]{1,4}"`
# some gnu installations gives off 'gcc version xx.xx.xx' instead of 'g++ version xx.xx.xx'                                   
if [ -z "$cxx_version_info" ]; then
  cxx_version_info=`$cxx_search -v 2>&1 | grep -oE "$cc_search\s[vV]ersion\s[0-9]{1,4}.[0-9]{1,4}.[0-9]{1,4}"`
fi
```
